### PR TITLE
Support objects in Schema.json, Form Validation, backend Debugger

### PR DESCRIPTION
**Pull request by zolga**
1. Improve jscode descripton + add @prop
2. Schema.json - Support reference to object: 
    tested via ajv-cli
3. Improve contact add/edit form validation:
    tested manually

### CODE REVIEW

File: backend/schema.json [hunk 1]
```diff
@@ -1,44 +1,36 @@
...
   "type": "object",
-  "title": "Phonebook contact root schema",
-  "description": "A contact from the phonebook",
-  "required": ["uid"],
   "properties": {
-    "uid": {
-      "$id": "#/properties/uid",
+    "contacts": {
+      "type": "object",
+      "description": "The unique identifier for a contact",
```
Comment (arielshaqed) [inline backend/schema.json]: This looks like a left-over from a previous object; a data `object` is usually a poor choice for a "unique identifier".

File: backend/schema.json [hunk 2]
```diff
@@ -1,44 +1,36 @@
...
+    }
+  },
+
+  "definitions": {
+    "contact": {
+      "$id": "#contact",
       "type": "object",
-      "title": "The unique identifier for a contact",
-      "required": ["uid", "name", "phone"],
+      "description": "The unique identifier for a contact",
```
Comment (arielshaqed) [inline backend/schema.json]: Here again.

File: backend/schema.json [hunk 3]
```diff
@@ -1,44 +1,36 @@
...
       "properties": {
         "uid": {
-          "$id": "#/properties/uid/properties/uid",
           "type": "integer",
-          "title": "The unique identifier for a contact",
-          "description": "The unique identifier for a contact",
-          "default": 0,
-          "examples": [1]
+          "description": "The unique identifier for a contact"
```
Comment (arielshaqed) [inline backend/schema.json]: This one seems correct.

File: backend/schema.json [hunk 4]
```diff
@@ -1,44 +1,36 @@
...
+    "contact": {
+      "$id": "#contact",
       "type": "object",
-      "title": "The unique identifier for a contact",
-      "required": ["uid", "name", "phone"],
+      "description": "The unique identifier for a contact",
       "properties": {
         "uid": {
-          "$id": "#/properties/uid/properties/uid",
           "type": "integer",
```
Comment (arielshaqed) [inline backend/schema.json]: `integer` is not a good type for a unique ID.  Even with JavaScript best-case assumptions you get ~53 bits in an integer.  Random selection of UIDs may start failing before even 10_000_000 different objects.  While that's a lot for a single user, as soon as you want to allow merging data you'll have a hard time.
Probably easiest to use type `string` and store either a UUID v4 or (nicer) a nanoid.

File: src/components/NavigationBar.js [hunk 5]
```diff
@@ -11,8 +11,11 @@ import phone from '../assets/phone.png';
 /**
  * Navigation bar of the app, hide/show the add new content button.
  *
- * @param {string} display - current page that is dispalyed.
- * @param {string} setDisplay - change the dispalyed page.
+ * @typedef {object} Props
+ * @prop {string} display - current page that is dispalyed
```
Comment (arielshaqed) [inline src/components/NavigationBar.js]: nit:
```suggestion
 * @prop {string} display - current page that is displayed
```
(also same change on next line)

File: src/components/form/ContactForm.js [hunk 6]
```diff
@@ -46,17 +56,19 @@ export default function ContactForm({
   const [formValid, setFormValid] = useState(name && phone);
 
   /* validator for name input */
-  function updateName(value) {
+  function updateName(value, event) {
     const isValid = value && 0 < value.trim().length;
     setName(isValid ? value : null);
     setFormValid(isValid && phone);
+    event && onSubmit(event);
```
Comment (arielshaqed) [inline src/components/form/ContactForm.js]: This is a very Perl-ish style.  Unless it's common in React, I think I prefer
```js
if (event) onSubmit(event);
```
(also below)

File: src/components/list/Contact.js [hunk 7]
```diff
@@ -9,12 +9,15 @@ import pen from '../../assets/pen.png';
...
-/*
+/**
  * Render a single phone book entry.
  *
- * @param {object} entry - phone book entry{name, phone}
- * @param {function} props.onEdit - trigger edit operation
- * @param {function} props.onRemove - delete item from the list
+ * @typedef {object} Props
+ * @prop {object} contact - phone book contact{name, phone}
+ * @prop {function} props.onEdit - trigger edit operation
```
Comment (arielshaqed) [inline src/components/list/Contact.js]: Why do the 2 functions get a prefix "`props.`"?

File: backend/schema.json (hunk 4 above)
Comment (zolga) [reply backend/schema.json]: An issue was created. Thanks!

File: src/components/NavigationBar.js (hunk 5 above)
Comment (zolga) [reply src/components/NavigationBar.js]: fixed. thanks!

File: src/components/list/Contact.js (hunk 7 above)
Comment (zolga) [reply src/components/list/Contact.js]: You are right, I removed it. thanks!

File: backend/schema.json [hunk 8]
```diff
@@ -1,44 +1,36 @@
...
-  "title": "Phonebook contact root schema",
-  "description": "A contact from the phonebook",
-  "required": ["uid"],
   "properties": {
-    "uid": {
-      "$id": "#/properties/uid",
+    "contacts": {
+      "type": "object",
+      "description": "The unique identifier for a contact",
+      "$ref": "#contact"
```
Comment (vogre) [inline backend/schema.json]:
```suggestion
      "additionalProperties": { "$ref": "#contact" }
```

Comment (zolga) [reply backend/schema.json]: Thanks! I changed by your suggestion.

**Comment by zolga** [comment]
@arielshaqed @vogre  PTAL

File: backend/schema.json (hunk 4 above)
Comment (arielshaqed) [reply backend/schema.json]: Thanks!
FTR that's #12.

**Comment by zolga** [comment]
@vogre @arielshaqed 
 I made changes to the schema.json with your suggestions.

File: backend/schema.json [hunk 9]
```diff
@@ -6,16 +6,20 @@
   "properties": {
     "contacts": {
       "type": "object",
-      "description": "The unique identifier for a contact",
+      "description": "The contacts hash",
       "additionalProperties": { "$ref": "#contact" }
+    },
+    "uid": {
+      "type": "integer",
+      "description": "The last allocated uid"
```
Comment (arielshaqed) [inline backend/schema.json]: 1. Please name it accordingly -- I expected a field "UID" to give a UID for its object.  Maybe `nextUid` or `lastUid` (depending on usage)?
2. Should this be required?
2. This probably emphasizes that we should move away from (a) sequential UIDs and (b) storing all contacts in a single Document.

Points (1) & (2) are for this PR, point (3) is *not* for this PR.

File: src/components/NavigationBar.js [hunk 10]
```diff
@@ -11,8 +11,11 @@ import phone from '../assets/phone.png';
 /**
  * Navigation bar of the app, hide/show the add new content button.
  *
- * @param {string} display - current page that is dispalyed.
- * @param {string} setDisplay - change the dispalyed page.
+ * @typedef {object} Props
+ * @prop {string} display - current page that is displayed
+ * @prop {string} setDisplay - change the displayed page
```
Comment (arielshaqed) [inline src/components/NavigationBar.js]: AFAIU this is a function not a string.

File: src/components/SubmitButton.js [hunk 11]
```diff
@@ -8,9 +8,13 @@ import '../style/submitButton.css';
  * Dynamic button that changed by updating buttons props from a state change.
  * This will gives the user feedback to the saving state.
  *
- * @param {string} inProcess - tracks whether the form is in process.
- * @param {string} handleClick - submit the form
- * @param {string} done - the form submited with successful result, disable the button to avoid resubmit
+ * @typedef {Object} Props
+ * @prop {string} inProcess - tracks whether the form is in process
```
Comment (arielshaqed) [inline src/components/SubmitButton.js]: Is this a string?

File: src/components/SubmitButton.js [hunk 12]
```diff
@@ -8,9 +8,13 @@ import '../style/submitButton.css';
...
+ * @prop {string} handleClick - submit the form
```
Comment (arielshaqed) [inline src/components/SubmitButton.js]: AFAIU this is a function not a string.

File: src/components/SubmitButton.js [hunk 13]
```diff
@@ -8,9 +8,13 @@ import '../style/submitButton.css';
...
+ * @prop {string} done - the form submited with successful result, disable the button to avoid resubmit
```
Comment (arielshaqed) [inline src/components/SubmitButton.js]: AFAIU this is a boolean not a string.

Also nit: "submit*ted".

File: src/components/SubmitButton.js [hunk 14]
```diff
@@ -8,9 +8,13 @@ import '../style/submitButton.css';
...
+ * @prop {boolean} disabled - enable/deisble the button
```
Comment (arielshaqed) [inline src/components/SubmitButton.js]:
```suggestion
 * @prop {boolean} disabled - enable/disable the button
```

File: src/components/form/FormField.js [hunk 15]
```diff
@@ -2,16 +2,19 @@ import React, { useState } from 'react';
...
- * @param {string} invalid - feedback for invalid values
- * @param {function} update - update parent with new value
- * @param {function} clearErrors - remove alert box messages when typing
- * @param {string} placeholder - field placeholder
+ * @typedef {object} Props
+ * @prop {string} label - field name
+ * @prop {string} value - initial value
+ * @prop {boolean} focus - should focus on initial render
+ * @prop {string} invalid - feedback for invalid values
+ * @prop {function} update - update parent with new value
```
Comment (arielshaqed) [inline src/components/form/FormField.js]: Thinking about it, the JSDoc type `function` is not particularly informative: how many parameters does it have?  What is the return type??
JSDoc seems to suggest using the [`@callback` tag](https://jsdoc.app/tags-callback.html).

File: src/components/form/ContactForm.js [hunk 16]
```diff
@@ -21,7 +21,17 @@ import {
   formType
 } from '../../utils/constants';
 
-/* Render a form to edit or create a phone book contact. */
+/**
+ * Render a form to edit or create a phone book contact.
+ *
+ * @typedef {object} Props
+ * @prop {string} display - the current page with uid if on edit mode
```
Comment (arielshaqed) [inline src/components/form/ContactForm.js]: AFAIU it isn't a string, e.g. in line 44 we access `display.uid`.

File: src/components/form/ContactForm.js (hunk 6 above)
Comment (arikmaor) [reply src/components/form/ContactForm.js]: it's common
also, I like it

File: backend/schema.json (hunk 9 above)
Comment (zolga) [reply backend/schema.json]: @arielshaqed in (1), you requesting change the description or the field name "uid":? 
basically, the backend is updating 2 documents: contacts & uid
uid document is used as a basic counter and keeps the last allocated uid.

about comment (2) - the default value is 0, (now I see that this fact is missing and I will add default property), and the requirement is not related to this uid.
I think there is a missing part and because of that there is confusion, please take a look at this addition under contacts object:

```
"properties": {
    "contacts": {
      "type": "object",
      "description": "The contacts hash",
      "properties": {
        "uid": {
          "type": "integer",
          "description": "The unique identifier for a contact",
          "additionalProperties": { "$ref": "#contact" }
        }
      },
      "required": ["uid"]
    },
    "uid": {
      "type": "integer",
      "default": 0,
      "description": "The last allocated uid"
    }
  },
```
any feedback will be appreciated!

File: src/components/form/FormField.js (hunk 15 above)
Comment (zolga) [reply src/components/form/FormField.js]: opened an issue #13

File: backend/schema.json (hunk 9 above)
Comment (arielshaqed) [reply backend/schema.json]: I understand.  I think the top-level property named `uid` should instead be named `lastUid`.  Even in this document, the field `uid` is a "unique identifer" of the object... except on the entire document, where it is the "last allocated UID" and not an identifier.  (You could even name it `lastAllocatedUid` and get rid of the documentation!)
Meanwhile the field `uid on an individual contact should remain `uid`: it is indeed "[t]he unique identifier for a contact".

Comment (zolga) [reply backend/schema.json]: @arielshaqed can you explain what do you mean by "get rid of the documentation"?

Comment (arielshaqed) [reply backend/schema.json]: Thanks for the change!

Note that the description "[t]he last allocated uid" now becomes unnecessary for a variable `lastAllocatedUid` (and some programming styles would omit it, at least from programs).

Anyway this change is good, thanks!

File: src/components/form/ContactForm.js [hunk 17]
```diff
@@ -19,9 +19,19 @@ import {
...
-} from '../../utils/constants';
+} from "../../utils/constants";
 
-/* Render a form to edit or create a phone book contact. */
+/**
+ * Render a form to edit or create a phone book contact.
+ *
+ * @typedef {object} Props
+ * @prop {object} display - the current page with uid if on edit mode
+ * @prop {object} contacts - list of all the contacts
```
Comment (arielshaqed) [inline src/components/form/ContactForm.js]: Type (and I think code below) says `object`, documentation says `list`.  I think they are not both correct.

File: src/components/form/ContactForm.js [hunk 18]
```diff
@@ -19,9 +19,19 @@ import {
...
+ * @prop {function} setDisplay - set the display
+ * @prop {function} setContacts - set a new list after changes: delete/edit/add
```
Comment (arielshaqed) [inline src/components/form/ContactForm.js]: As above: I don't think `setContacts` takes a list.

File: src/components/list/Contact.js [hunk 19]
```diff
@@ -9,12 +9,15 @@ import pen from '../../assets/pen.png';
...
 
-/*
+/**
  * Render a single phone book entry.
  *
- * @param {object} entry - phone book entry{name, phone}
- * @param {function} props.onEdit - trigger edit operation
- * @param {function} props.onRemove - delete item from the list
+ * @typedef {object} Props
+ * @prop {object} contact - phone book contact{name, phone}
```
Comment (arielshaqed) [inline src/components/list/Contact.js]: nit:
```suggestion
 * @prop {object} contact - phone book contact name, phone}
```

File: backend/contacts.js [hunk 20]
```diff
@@ -8,7 +8,7 @@ import { MessageType } from './enums';
  */
```
Comment (arielshaqed) [inline backend/contacts.js]: (Also below) type is `object`, documentation says "list".  I think one of these is wrong.

File: backend/contacts.js [hunk 21]
```diff
@@ -1,5 +1,5 @@
-import { get, update } from '@binaris/shift-db';
-import { MessageType } from './enums';
+import { get, update } from "@binaris/shift-db";
```
Comment (arielshaqed) [inline backend/contacts.js]: Why this change?  I guess it's OK if it's consistent across our examples, but it seems odd.

In any case, if we have a decision how we want to do it, please add (in a separate PR) linting: humans aren't good at following such rules and we need a computer to keep them in line.

File: src/components/SubmitButton.js (hunk 11 above)
Comment (arielshaqed) [reply src/components/SubmitButton.js]: I think this is wrong.  [L. 188](https://github.com/shiftjs/phonebook/pull/11/files#diff-9ac25535f022691f9da0d2dfdcbee9f3R188) seems to pass a boolean into `inProcess`, and this component only ever looks at the truthiness of `inProcess`.

Comment (arielshaqed) [reply src/components/SubmitButton.js]: I believe this is still incorrect.
